File: Android/scenes/ktv/src/main/java/io/agora/scene/ktv/widget/song/SongDialog.java
package io.agora.scene.ktv.widget.song;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.fragment.app.Fragment;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;

import java.util.List;

import io.agora.scene.base.component.BaseBottomSheetDialogFragment;
import io.agora.scene.ktv.R;
import io.agora.scene.ktv.databinding.KtvDialogChooseSongBinding;
import io.agora.scene.ktv.live.listener.SongActionListenerImpl;
import io.agora.scene.widget.utils.UiUtils;

/**
 * Song menu
 */
public class SongDialog extends BaseBottomSheetDialogFragment<KtvDialogChooseSongBinding> {

    private OnSongActionListener chooseSongListener;

    private final SongChosenFragment songChosenFragment = new SongChosenFragment();
    private final SongChooseFragment songChooseFragment = new SongChooseFragment();

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ViewCompat.setOnApplyWindowInsetsListener(requireDialog().getWindow().getDecorView(), (v, insets) -> {
            Insets inset = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            mBinding.pager.setPadding(0, 0, 0, inset.bottom);
            return WindowInsetsCompat.CONSUMED;
        });
        mBinding.rBtnChooseSong.setChecked(true);
        mBinding.pager.getChildAt(0).setOverScrollMode(View.OVER_SCROLL_NEVER);

        songChooseFragment.setListener(new SongChooseFragment.Listener() {
            @Override
            public void onSongItemChosen(@NonNull SongItem songItem) {
                if (UiUtils.isFastClick(500)) {
                    return;
                }
                if (chooseSongListener != null) {
                    chooseSongListener.onChooseSongChosen(SongDialog.this, songItem);
                }
            }

            @Override
            public void onSongsSearching(String condition) {
                if (chooseSongListener != null) {
                    chooseSongListener.onChooseSongSearching(SongDialog.this, condition);
                }
            }

            @Override
            public void onSongsRefreshing() {
                if (chooseSongListener != null) {
                    chooseSongListener.onChooseSongRefreshing(SongDialog.this,0);
                }
            }

            @Override
            public void onSongsLoadMore() {
                if (chooseSongListener != null) {
                    chooseSongListener.onChooseSongLoadMore(SongDialog.this, 0);
                }
            }
        });
        songChosenFragment.setListener(new SongChosenFragment.Listener() {
            @Override
            public void onSongDeleteClicked(SongItem song) {
                if (chooseSongListener != null) {
                    chooseSongListener.onChosenSongDeleteClicked(SongDialog.this, song);
                }
            }

            @Override
            public void onSongTopClicked(SongItem song) {
                if (chooseSongListener != null) {
                    chooseSongListener.onChosenSongTopClicked(SongDialog.this, song);
                }
            }
        });
        Fragment[] fragments = new Fragment[]{songChooseFragment, songChosenFragment};
        mBinding.pager.setSaveEnabled(false);
        mBinding.pager.setAdapter(new FragmentStateAdapter(getChildFragmentManager(), getViewLifecycleOwner().getLifecycle()) {
            @Override
            public int getItemCount() {
                return fragments.length;
            }

            @NonNull
            @Override
            public Fragment createFragment(int position) {
                return fragments[position];
            }
        });
        mBinding.pager.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {

            @Override
            public void onPageSelected(int position) {
                super.onPageSelected(position);
                if (position == 0) {
                    mBinding.rBtnChooseSong.setChecked(true);
                } else {
                    mBinding.rBtnChorus.setChecked(true);
                }
            }
        });
        setChosenSongCount(0);
    }

    @Override
    public void onStart() {
        super.onStart();
        mBinding.radioGroup.setOnCheckedChangeListener((radioGroup, i) -> {
            if (i == R.id.rBtnChooseSong) {
                mBinding.pager.setCurrentItem(0);
            } else if(i == R.id.rBtnChorus){
                mBinding.pager.setCurrentItem(1);
            }
        });
    }

    /**
     *  set event listener
     * @param chooseSongListener
     */
    public void setChooseSongListener(SongActionListenerImpl chooseSongListener) {
        this.chooseSongListener = chooseSongListener;
        chooseSongListener.getSongTypeList();
    }

    /**
     * Order - update item selected state.
     *
     * @param songItem the song item
     * @param isChosen the is chosen
     */
    public void setChooseSongItemStatus(SongItem songItem, boolean isChosen) {
        songChooseFragment.setSongItemStatus(songItem, isChosen);
    }

    /**
     * 点歌-更新搜索列表
     *
     * @param list the list
     */
    public void setChooseSearchResult(List<SongItem> list) {
        songChooseFragment.setSearchResult(list);
    }

    /**
     * 点歌-下拉刷新重置列表
     *
     * @param list  the list
     * @param index the index
     */
    public void setChooseRefreshingResult(List<SongItem> list,int index) {
        songChooseFragment.setRefreshingResult(list);
    }

    /**
     * 点歌-加载更多刷新列表
     *
     * @param list    the list
     * @param hasMore the has more
     * @param index   the index
     */
    public void setChooseLoadMoreResult(List<SongItem> list, boolean hasMore, int index) {
        songChooseFragment.setLoadMoreResult(list, hasMore);
    }

    /**
     * Queued song list - set whether actions like delete or pin to top are allowed.
     *
     * @param controllable the controllable
     */
    public void setChosenControllable(boolean controllable) {
        songChosenFragment.setControllable(controllable);
    }

    /**
     * Queued song list - reset the list.
     *
     * @param songs the songs
     */
    public void resetChosenSongList(List<SongItem> songs) {
        songChosenFragment.resetSongList(songs);
        setChosenSongCount(songChosenFragment.getSongSize());

        songChooseFragment.setRestSongStatus(songs);
    }

    /**
     *  Queued song list -add song item.
     *
     * @param song the song
     */
    public void addChosenSongItem(SongItem song) {
        songChosenFragment.addSongItem(song);
        setChosenSongCount(songChosenFragment.getSongSize());
    }

    /**
     *  Queued song list - delete song item.
     *
     * @param song the song
     */
    public void deleteChosenSongItem(SongItem song) {
        songChosenFragment.deleteSongItem(song);
        setChosenSongCount(songChosenFragment.getSongSize());
    }

    /**
     *  Queued song list - pin song item.
     *
     * @param song the song
     */
    public void topUpChosenSongItem(SongItem song) {
        songChosenFragment.topUpSongItem(song);
    }

    private void setChosenSongCount(int count) {
        if(mBinding == null){
            return;
        }
        if (count > 0) {
            mBinding.tvChoosedSongCount.setVisibility(View.VISIBLE);
            if (count > 99) {
                count = 99;
            }
            mBinding.tvChoosedSongCount.setText(String.valueOf(count));
        } else {
            mBinding.tvChoosedSongCount.setVisibility(View.GONE);
        }
    }

}

File: Android/scenes/cantata/src/main/java/io/agora/scene/cantata/widget/song/SongDialog.kt
package io.agora.scene.cantata.widget.song

import android.os.Bundle
import android.view.View
import android.widget.RadioGroup
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.fragment.app.Fragment
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2.OnPageChangeCallback
import com.scwang.smart.refresh.layout.api.RefreshLayout
import io.agora.scene.base.component.BaseBottomSheetDialogFragment
import io.agora.scene.cantata.R
import io.agora.scene.cantata.databinding.CantataDialogChooseSongBinding
import io.agora.scene.cantata.live.listener.SongActionListenerImpl
import io.agora.scene.widget.utils.UiUtils

/**
 * Song menu
 */
class SongDialog : BaseBottomSheetDialogFragment<CantataDialogChooseSongBinding?>() {
    private var chooseSongListener: OnSongActionListener? = null
    private val songChosenFragment = SongChosenFragment()
    private val songChooseFragment = SongChooseFragment()
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        ViewCompat.setOnApplyWindowInsetsListener(requireDialog().window!!.decorView) { v: View?, insets: WindowInsetsCompat ->
            val inset = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            mBinding?.pager?.setPadding(0, 0, 0, inset.bottom)
            WindowInsetsCompat.CONSUMED
        }
        mBinding?.apply {
            rBtnChooseSong.isChecked = true
            pager.getChildAt(0)?.overScrollMode = View.OVER_SCROLL_NEVER
            songChooseFragment.listener = object : SongChooseFragment.Listener {
                override fun onClickSongItem(songItem: SongItem) {
                    if (UiUtils.isFastClick(500)) return
                    chooseSongListener?.onChooseSongChosen(this@SongDialog, songItem)
                }


                override fun onRefresh(refreshLayout: RefreshLayout) {
                    chooseSongListener?.onChooseSongRefreshing(this@SongDialog)
                }

                override fun onLoadMore(refreshLayout: RefreshLayout) {
                }
            }
            songChosenFragment.setListener(object : SongChosenFragment.Listener {
                override fun onSongDeleteClicked(song: SongItem?) {
                    chooseSongListener?.onChosenSongDeleteClicked(this@SongDialog, song!!)
                }

                override fun onSongTopClicked(song: SongItem?) {
                    chooseSongListener?.onChosenSongTopClicked(this@SongDialog, song!!)
                }
            })
            val fragments = arrayOf<Fragment>(songChooseFragment, songChosenFragment)
            pager.isSaveEnabled = false
            pager.adapter = object : FragmentStateAdapter(childFragmentManager, viewLifecycleOwner.lifecycle) {
                override fun getItemCount(): Int {
                    return fragments.size
                }

                override fun createFragment(position: Int): Fragment {
                    return fragments[position]
                }
            }
            pager.registerOnPageChangeCallback(object : OnPageChangeCallback() {
                override fun onPageSelected(position: Int) {
                    super.onPageSelected(position)
                    if (position == 0) {
                        rBtnChooseSong.isChecked = true
                    } else {
                        rBtnChorus.isChecked = true
                    }
                }
            })
            setChosenSongCount(0)
        }

    }

    override fun onStart() {
        super.onStart()
        mBinding?.radioGroup?.setOnCheckedChangeListener { radioGroup: RadioGroup?, i: Int ->
            if (i == R.id.rBtnChooseSong) {
                mBinding?.pager?.currentItem = 0
            } else if (i == R.id.rBtnChorus) {
                mBinding?.pager?.currentItem = 1
            }
        }
    }

    /**
     * Set event listener
     */
    fun setChooseSongListener(chooseSongListener: SongActionListenerImpl) {
        this.chooseSongListener = chooseSongListener
    }


    /**
     * Song selection - Update item selected status
     */
    fun setChooseSongItemStatus(songItem: SongItem, isChosen: Boolean) {
        songChooseFragment.setSongItemStatus(songItem, isChosen)
    }


    /**
     * Song selection - Reset list when pull-to-refresh
     */
    fun setChooseRefreshingResult(list: List<SongItem>) {
        songChooseFragment.setRefreshingResult(list)
    }


    /**
     * Selected song list - Set whether to delete, top, etc.
     */
    fun setChosenControllable(controllable: Boolean) {
        songChosenFragment.setControllable(controllable)
    }

    /**
     * Selected song list - Reset list
     */
    fun resetChosenSongList(songs: List<SongItem?>?) {
        songChosenFragment.resetSongList(songs)
        setChosenSongCount(songChosenFragment.songSize)
    }

    /**
     * Selected song list - Add song
     */
    fun addChosenSongItem(song: SongItem?) {
        songChosenFragment.addSongItem(song)
        setChosenSongCount(songChosenFragment.songSize)
    }

    /**
     * Selected song list - Delete song
     */
    fun deleteChosenSongItem(song: SongItem?) {
        songChosenFragment.deleteSongItem(song!!)
        setChosenSongCount(songChosenFragment.songSize)
    }

    /**
     * Selected song list - Top song
     */
    fun topUpChosenSongItem(song: SongItem?) {
        songChosenFragment.topUpSongItem(song!!)
    }

    private fun setChosenSongCount(count: Int) {
        var localCount = count
        mBinding?.apply {
            if (count > 0) {
                tvChoosedSongCount.visibility = View.VISIBLE
                if (localCount > 99) {
                    localCount = 99
                }
                tvChoosedSongCount.text = count.toString()
            } else {
                tvChoosedSongCount.visibility = View.GONE
            }
        }
    }
}